dbms/src/Storages/KVStore/Utils/AsyncTasks.h

```diff
@@ -14,110 +14,358 @@
 
 #pragma once
 
+#include <Common/FmtUtils.h>
+#include <Common/Logger.h>
+#include <Common/SyncPoint/SyncPoint.h>
 #include <Common/UniThreadPool.h>
 
 #include <future>
+#include <magic_enum.hpp>
+#include <unordered_map>
 
 namespace DB
 {
+
+namespace AsyncTaskHelper
+{
+// It's guaranteed a task is no longer accessible once canceled or has its result fetched.
+// NotScheduled -> InQueue, Running, NotScheduled
+// InQueue -> Running
+// Running -> Finished, NotScheduled(canceled)
+// Finished -> NotScheduled(fetched)
+// NOTE: magic_enum can't work properly with enums in template classes due to compiler opts.
+enum class TaskState
+{
+    NotScheduled,
+    InQueue,
+    Running,
+    Finished,
+};
+} // namespace AsyncTaskHelper
+
 template <typename Key, typename Func, typename R>
 struct AsyncTasks
 {
     // We use a big queue to cache, to reduce add task failures.
     explicit AsyncTasks(uint64_t pool_size, uint64_t free_pool_size, uint64_t queue_size)
         : thread_pool(std::make_unique<ThreadPool>(pool_size, free_pool_size, queue_size))
+        , log(DB::Logger::get("AsyncTasks"))
     {}
 
-    bool discardTask(Key k)
+    ~AsyncTasks() { LOG_INFO(log, "Pending {} tasks when destructing", count()); }
+
+    using TaskState = AsyncTaskHelper::TaskState;
+
+    struct CancelHandle;
+    using CancelHandlePtr = std::shared_ptr<CancelHandle>;
+    struct CancelHandle
+    {
+        CancelHandle() = default;
+        CancelHandle(const CancelHandle &) = delete;
+
+        bool canceled() const { return inner.load(); }
```

Comment:
```suggestion
        bool isCanceled() const { return inner.load(); }
```

Reply:
Why? `canceled` already means a state rather than a operation.

Reply:
It is hard to tell the difference between action `cancel` and status `canceled` at one glance. I used to think there is a public method of "action cancel" for this handle class. Actually the action is a private method can only be called by `AsyncTasks`.

`isCanceled` is more explicit